### PR TITLE
Move ledger translation from ouroboros-consensus

### PR DESCRIPTION
**Pull request by kderme**
https://github.com/input-output-hk/ouroboros-network/issues/2253

### CODE REVIEW

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/HardFork.hs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+import Shelley.Spec.Ledger.Rewards
+import Shelley.Spec.Ledger.Slot
+import Shelley.Spec.Ledger.UTxO
+
+-- | We use the same hashing algorithm so we can unwrap and rewrap the bytes.
+-- We don't care about the type that is hashed, which will differ going from
+-- Byron to Shelley, we just use the hashes as IDs.
+translateTxIdByronToShelley :: Crypto sc => Byron.TxId -> TxId sc
+translateTxIdByronToShelley =
+    TxId . hashFromBytesE . Hashing.hashToBytes
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/HardFork.hs]: Actually, we can improve this (possibly in a separate commit): both types are `ShortByteString`s under the hood, so we should avoid converting via a `ByteString`.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/HardFork.hs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        , _pstate = emptyPState
+        }
+      }
+    oSchedule :: OverlaySchedule sc
+    oSchedule =
+      flip runReader globals $
+        overlaySchedule
+          epochNo
+          (Map.keysSet (sgGenDelegs genesisShelley))
+          (sgProtocolParams genesisShelley)
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/HardFork.hs]: Also incorporate the comments on the issue, e.g., https://github.com/input-output-hk/ouroboros-network/issues/2253#issuecomment-654842803.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/HardFork.hs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+-- length.
+hashFromBytesE
+  :: forall h a. (Byron.HashAlgorithm h, HasCallStack)
+  => Strict.ByteString
+  -> Byron.Hash h a
+hashFromBytesE bs = fromMaybe (error msg) $ Byron.hashFromBytes bs
+  where
+    msg =
+      "hashFromBytes called with ByteString of the wrong length: " <>
+      show bs
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/HardFork.hs]: I would move it closer to `translateTxIdByronToShelley`.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/HardFork.hs [hunk 4]
```diff
@@ -0,0 +1,171 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Shelley.Spec.Ledger.HardFork
+  ( mkInitialShelleyLedgerView,
+    translateToShelleyLedgerState,
+      -- * Exported for testing purposes
+    translateTxIdByronToShelley,
+    translateCompactTxOutByronToShelley
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/HardFork.hs]: You should move the `Test.Consensus.Cardano.Translation` test module as well.

**Review by mrBliss — COMMENTED** [review]
I would rename this to PR to "Move ledger translation from ouroboros-consensus". I would also rename the module to `Translation.hs` or `ByronTranslation.hs`. The ledger doesn't really have to be aware of hard forks.

BTW, you can open a PR as a draft instead of having "WIP" as the description.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/ByronTranslation.hs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+import Shelley.Spec.Ledger.LedgerState
+import Shelley.Spec.Ledger.OverlaySchedule
+import Shelley.Spec.Ledger.Rewards
+import Shelley.Spec.Ledger.Slot
+import Shelley.Spec.Ledger.UTxO
+
+-- | We use the same hashing algorithm so we can unwrap and rewrap the bytes.
+-- We don't care about the type that is hashed, which will differ going from
+-- Byron to Shelley, we just use the hashes as IDs.
+translateTxIdByronToShelley :: Crypto sc => Byron.TxId -> TxId sc
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/ByronTranslation.hs]: We can add a `ADDRHASH c ~ Blake2b_224` constraint, because we rely on that.

**Review by nc6 — APPROVED** [review]

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/ByronTranslation.hs (hunk 5 above)
Comment (mrBliss) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/ByronTranslation.hs]: We should do this before merging.

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/ByronTranslation.hs [hunk 6]
```diff
@@ -0,0 +1,190 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
```
Comment (mrBliss) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/ByronTranslation.hs:4]: Why is this needed?

Comment (kderme) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/ByronTranslation.hs:4]: For the constraint you mentioned above `ADDRHASH c ~ Blake2b_224`

Comment (mrBliss) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/ByronTranslation.hs:4]: Oh :facepalm:. I thought it was maybe related to a redundant `HasCallStack` (I have seen such constraints in this repo before).

Comment (mrBliss) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/ByronTranslation.hs:4]: (BTW, we have https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-consensus/src/Ouroboros/Consensus/Util/RedundantConstraints.hs for that in consensus, but not here of course)

**Review by mrBliss — APPROVED** [review]
